### ISSUE TITLE
podman storage usage with default (vfs) driver "eats" a lot of space

### ISSUE THREAD

**Issue by vneno** [unifi-utilities/unifios-utilities#492]
**Describe the bug**
podman storage usage with default (vfs) driver "eats" a lot of space - much more than it should.
It seems that this issue is realted to images with multiple blobs.

**To Reproduce**
1. check storage usage : du -hs /var/lib/containers/storage/vfs; podman system df; df -h /
2. pull some image: i.e.: podman  pull docker.io/pihole/pihole:latest
3. check storage usage again: du -hs /var/lib/containers/storage/vfs; podman system df; df -h /
4. compare storage usage before/after pull.

**Screenshots**
Before pull:
# du -hs /var/lib/containers/storage/vfs
1.9G	/var/lib/containers/storage/vfs
# podman system df
TYPE           TOTAL       ACTIVE      SIZE        RECLAIMABLE
Images         5           4           907.8MB     112.5MB (0%)
Containers     4           3           52.28MB     732.2kB (0%)
Local Volumes  5           0           0B          0B (0%)

After pull:
# du -hs /var/lib/containers/storage/vfs
4.1G	/var/lib/containers/storage/vfs
# podman system df
TYPE           TOTAL       ACTIVE      SIZE        RECLAIMABLE
Images         6           4           1.23GB      434.7MB (0%)
Containers     4           3           52.2MB      732.2kB (0%)
Local Volumes  5           0           0B          0B (0%)

**UDM Information**
 Model: UniFi Dream Machine (UDM)
 Firmware: 2.4.27 (2.4.27)
 podman version 3.4.4
 
**Additional context**
This issue is relevant to the udmpro and udmse podman.

**Workaround**
Pull the image, create container, export it as a tamplate to tar file, then create final container as import from that tar.
Example: 
podman pull docker.io/pihole/pihole:latest
podman run --name  pihole-template --entrypoint /bin/date docker.io/pihole/pihole:latest
podman export pihole-template > /persistent/custom/pihole.tar
podman rm -f pihole-template
podman image rm -f docker.io/pihole/pihole:latest
podman import /persistent/custom/pihole.tar pihole

**Comment by rodalpho** [comment]
I've run into this problem before running docker in a LXC. The correct fix is to install fuse-overlayfs, basically the default overlayfs but running in userspace. After that docker will automatically detect it's available,  stop using vfs, and stop taking up all your disk space.

Note your other fix is likely going to keep using tons of space as your image updates. I guess it's fine if you never update the image. 

The above was my experience running docker running nested in a LXC, and also needed a static fuse-overlayfs inside the LXC, while in this issue we're talking about podman running on bare metal ARM, but I imagine it should all transfer over. 

TL;DR: Try fuse-overlayfs.

**Comment by vneno** [comment]
Yes, the fuse-overlayfs is probably the best solution, but the UDM base OS is based on Debian 9.13 and the package is available since buster. The fuse kernel module is also needed, and UDM kernel seems to be customized "4.19.152-ui-alpine-udm" so building specific kernel module would be necessary also.

The workaround  is just a temporarly (as I hope) solution. But updating images this way takes only few more minutes than usual pull, but use more resources (CPU,RAM) I've created scripts to update images of containers I'm using. And I run them manually, so far so good.

**Comment by meaterator** [comment]
@vneno 

Really glad you posted this. Thought I was going crazy. Came here to report this tonight. Thanks for the workaround. Installing a simple 1GB image eats up almost 6GB, which is all the free space I have.

**Comment by meaterator** [comment]
For anyone trying to do the above workaround for homebridge, note that because you're exporting/importing a running container vs. saving/loading an image, you are going to lose some arguments that would normally get passed along, most importantly, the entrypoint. 

There are a number of ways to get an entrypoint, and a number of ways to pass an entrypoint. If you just want the entrypoint for oznu's homebridge, it's **"/init"**. You need to pass it without the --entrypoint flag though. Just put it at the end of your run command, with quotes. If you're going to pass it with --entrypoint you'll need to put it in brackets to be passed as JSON. Known issue with how podman handles entrypoints vs. docker.

**Comment by boostchicken** [comment]
I have tried evverything I can to get fuse-overlayfs to work, now podman builds have an overlay driver also, try them.  It just doesnt work evevn though there is an ovverlay rootfs.  Hopefully thiis isnt an issue in 3 and if it there will be better support for fuse because we won't be on a way out dated kernel.

**Comment by boostchicken** [comment]
@vneno could I trouble you to write a macro so that if people want to make containers they just the stuff usally would, but to your script.

**Comment by boostchicken** [comment]
Also I have the 3 kernel source IO am still waiting on 2.5 Ubi is so slow.

**Comment by vneno** [comment]
@boostchicken thanks for the hint, but /persistant partition is small in udm (2GB)
I think for UDM this should work fine:

1. change storage driver in /data/podman/conf/storage.conf to: driver = "overlay"
2. move containers dir out of the overlay: mv /var/lib/containers /mnt/.rwfs/
3. create symlink back to the original path: ln -s /mnt/.rwfs/containers /var/lib/ 

Now everithing works as before 2.4 upgrade and it survived reboot also.

**Comment by MRobi1** [comment]
After this workaround, how do you run the container using the saved image file?
if I run podman images I can see the image as localhost
But trying to run it using the repository name, image name or image id I just get
Error: no command or entrypoint provided, and no CMD or ENTRYPOINT from image

And of course if I use the original command it just fills my storage right back up

**Comment by vneno** [comment]
A better "workaround" which for this
https://github.com/orgs/unifi-utilities/discussions/471#discussioncomment-5002838

for my UDM, this works:
- create dir on persistent storage - /mnt/.rwfs/containers (mkdir /mnt/.rwfs/containers)
- stop all containers (podman stop XXX)
- mv containers data to /mnt/.rwfs/containers (mv /var/lib/containers/* /mnt/.rwfs/containers/)
- create on_boot.d script which should start before any podman container
- this script shold bind or link the path /mnt/.rwfs/containers to /var/lib/containers
- run/start/pull container/image


My onboot script - if the path /var/lib/containers is a link or dir. If it is a dir then move it and create a link.

00-podman-relink.sh:
#!/bin/bash
if [ ! -s /var/lib/containers ]; then
    echo "not a link" 
    mv /var/lib/containers /var/lib/containers_disabled_$(date +%Y-%m-%d-%H%M)
    ln -sf /mnt/.rwfs/containers /var/lib/
fi

For UDMP/SE the approach should be similar but maybe with different target storage path.

For now i am done with UDM as it has low RAM available for custom usage, so I've moved to pfsense on intel NUC.

**Comment by MRobi1** [comment]
I assume this is because the specified path is overlayfs instead of vfs?
```
/dev/disk/by-partlabel/overlay      11G  2.6G  7.1G  27% /mnt/.rwfs
overlayfs-root                      11G  2.6G  7.1G  27% /
```

If so, is there a reason podman would still be creating a VFS directory even after creating the link?
Containers are now being stored at /mnt/.rwfs/containers/storage/vfs/dir

And when trying to pull the image for nginx-proxy-manager with no other containers up I'm still running up against
`Error: writing blob: adding layer with blob "sha256:1ff6ef28a654a753b95a9033bbb51f28bfcef426b7c5488033f397d9b9c62743": error creating layer with ID "4d5c4d48fc5d9675e8e5fdf49968489c5f18e832b3addcf5ab94652fb5cf02b2": no space left on device`

I think my best bet here may be to pickup a drive that the UDMP will actually recognize and move it onto there instead

**Comment by MRobi1** [comment]
I found a 5TB HDD that I wasn't using and that the UDMP would actually recognize and mount. So I moved storage onto /volume1

Running 4 containers:
NGINX PM (Image Size: 845Mb) 
Cloudflare-ddns (Image Size: 74.5Mb) 
Authelia (Image Size: 55.3Mb) 
Adguardhome (Image Size: 66MB)

These 4 are taking just over 15Gb in storage on that drive for roughly 1Gb worth of images.

**Comment by dvcrn** [comment]
@vneno how did you get fuse-overlayfs running on UDM? I have the same issue, and running homebridge + hap is already eating 7GB with vfs and clogs up all the space on the UDM
